### PR TITLE
Wrap current domain API

### PR DESCRIPTION
**Pull request by kounelisagis**
Wrapping https://github.com/TileDB-Inc/TileDB/pull/5056

### CODE REVIEW

**Review by johnkerl — CHANGES_REQUESTED** [review]
@kounelisagis thanks!

Some things here look a bit odd. I wrote an array using the core public C++ API and am reading it using:

* core `2.25.0-rc1` built and installed in my sandbox
* This branch locally built using `pip install -e .` from within a `TileDB-Py` checkout

These look good:

```
>>> A = tiledb.open('snda')

>>> cd = A.schema.current_domain()

>>> cd.is_empty
False
```

This does not:
```
>>> cd.type
TypeError: Unregistered type : tiledb_current_domain_type_t

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/git/TileDB-Inc/TileDB-Py/tiledb/current_domain.py", line 29, in type
    return self._type()
TypeError: Unable to convert function return value to a Python type! The signature was
	(arg0: tiledb.cc.CurrentDomain) -> tiledb_current_domain_type_t
```

Personally, I'd love it if we could have a `__str__` for `NDRectangle` objects:

```
>>> cd.ndrectangle()
<tiledb.ndrectangle.NDRectangle object at 0x70f71328c180>
```

I'm not succeeding in accessing any of the ranges:

```
>>> nd = cd.ndrectangle()

>>> nd.range(0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/git/TileDB-Inc/TileDB-Py/tiledb/ndrectangle.py", line 57, in range
    raise tiledb.TileDBError(
tiledb.cc.TileDBError: Trying to get a range for an index out of bounds is not possible

>>> nd.range(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/git/TileDB-Inc/TileDB-Py/tiledb/ndrectangle.py", line 57, in range
    raise tiledb.TileDBError(
tiledb.cc.TileDBError: Trying to get a range for an index out of bounds is not possible

>>> nd.range('soma_dim_0')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/git/TileDB-Inc/TileDB-Py/tiledb/ndrectangle.py", line 61, in range
    raise tiledb.TileDBError(
tiledb.cc.TileDBError: Cannot get dimension index; Invalid dimension name

>>> nd.range('soma_dim_1')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/git/TileDB-Inc/TileDB-Py/tiledb/ndrectangle.py", line 61, in range
    raise tiledb.TileDBError(
tiledb.cc.TileDBError: Cannot get dimension index; Invalid dimension name
```

File: tiledb/array_schema.py [hunk 1]
```diff
@@ -387,6 +388,29 @@ def has_dim_label(self, name: str) -> bool:
         """
         return self._has_dim_label(self._ctx, name)
 
+    def add_enumeration(self, enmr):
```
[the remark this answers is not in the document]
Comment (johnkerl) [reply tiledb/array_schema.py:391]: For bonus I'd love it if `current_domain` were a non-parenthesized accessor:

* `A = tiledb.open(uri)`
* `A.schema` -- has no parentheses
* `A.schema.current_domain()` -- requires parentheses after the second `.`

File: tiledb/tests/test_current_domain.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        schema.set_current_domain(current_domain)
+
+        cd = schema.current_domain()
+        self.assertFalse(cd.is_empty)
+
+        self.assertEqual(cd.ndrectangle().range(0), range_one)
+
+        with self.assertRaises(tiledb.TileDBError):
+            cd.ndrectangle().range(1)
+
```
Comment (johnkerl) [inline tiledb/tests/test_current_domain.py:102]: Missing test cases for reading back arrays _already created_.

* Create an array at a tempfile URI
* Write it with current domain
  * ints case
  * strings case
* Close the array
* Open the array from the URI
* Get its domain and current domain
* Do assertions on the current domain which was read back from storage (not just the one at creation time, which you have now, and which is well-tested)

File: tiledb/ndrectangle.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        :param start: Range start value
+        :param end: Range end value
+        :raises tiledb.TileDBError:
+        """
+        # Set types only if range is set successfully
+        try:
+            self._set_range(dim, start, end)
+            self._dim_types[dim] = type(start).__name__
+        except Exception as e:
+            raise e
```
Comment (nguyenv) [inline tiledb/ndrectangle.py:44]: The try-except block here is unnecessary as we don't add additional handling or information to the exception.

File: tiledb/ndrectangle.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        """Gets the range for the given dimension.
+
+        :param dim: Dimension name or index
+        :return: Range as a tuple (start, end)
+        :raises tiledb.TileDBError:
+        """
+        if dim not in self._dim_types:
+            if isinstance(dim, int):
+                raise tiledb.TileDBError(
+                    "Trying to get a range for an index out of bounds is not possible"
```
Comment (nguyenv) [inline tiledb/ndrectangle.py:58]: Is this the correct error message? This is checking if the `dim` is an `int` but the error message indiciates a OOB error.

File: tiledb/tests/test_current_domain.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        se = tiledb.ArraySchemaEvolution(ctx)
+        se.expand_current_domain(new_current_domain)
+        se.array_evolve(uri)
+
+        array = tiledb.Array(uri, mode="r")
+        s = array.schema
+        cd = s.current_domain()
+        n = cd.ndrectangle()
+        self.assertEqual(n.range(0), range_two)
```
Comment (nguyenv) [inline tiledb/tests/test_current_domain.py:139]: Need to `array.close()` or use `with` when opening.

**Comment by johnkerl** [comment]
Here's something puzzling.

I was wondering how
https://github.com/TileDB-Inc/TileDB-Py/blob/265f84c81e4840401aa5b65956fe3a9e1c62b6bc/tiledb/tests/test_current_domain.py#L135-L139
could pass -- when I try something like that I get

```
>>> A.schema.current_domain().ndrectangle().range(0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/git/TileDB-Inc/TileDB-Py/tiledb/ndrectangle.py", line 57, in range
    raise tiledb.TileDBError(
tiledb.cc.TileDBError: Trying to get a range for an index out of bounds is not possible
```
I made this pair:

* `test-1.py` https://gist.github.com/johnkerl/8078ae2cd143954569926f8f001bb7d9  -- crashes like mine above
* `test-2.py` https://gist.github.com/johnkerl/c3182d358457216a3512a7402c830d23 -- does not crash

Surprisingly, the one-line change

```
$ diff test-1.py test-2.py
16c16
< #ndrect.set_range(0, range_one[0], range_one[1])
---
> ndrect.set_range(0, range_one[0], range_one[1])
```

which makes `test-2.py` pass and `test-1.py` fail is whether or not something is done to the local `ndrect` which ought to have nothing to do with the S3 storage of the array I'm asking for its current domain ...

File: tiledb/cc/current_domain.cc [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+            if (type_name == "str") {
+              return ndrect.range<std::string>(dim_name);
+            } else if (type_name == "int") {
+              return ndrect.range<int64_t>(dim_name);
+            } else if (type_name == "float") {
+              return ndrect.range<double>(dim_name);
+            } else {
+              TPY_ERROR_LOC("Unsupported type for NDRectangle's range");
+            }
+          })
```
Comment (johnkerl) [inline tiledb/cc/current_domain.cc:134]: Unclear why we'd restrict to only  `int64_t`, `double`, and `std::string`

File: tiledb/cc/current_domain.cc [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+      .def(
+          "_range",
+          [](NDRectangle &ndrect, unsigned dim_idx,
+             const std::string &type_name)
+              -> std::variant<std::array<double, 2>, std::array<std::string, 2>,
+                              std::array<int64_t, 2>> {
+            if (type_name == "str") {
```
Comment (johnkerl) [inline tiledb/cc/current_domain.cc:141]: Ditto here

**Comment by kounelisagis** [comment]
@nguyenv, @johnkerl thank you very much for your comments; they were really valuable!
Please take your time to check my latest changes. I hope all your issues are resolved.

For each dimension, we need its `dtype` so we know it when calling `NDRectangle::range`. However, since there is no direct access to this, the workaround you see is needed.

I change the tests, I hope they are more clear and cover every case a user would try.

**Comment by johnkerl** [comment]
Thanks so much @kounelisagis -- taking a look now! :)

**Comment by johnkerl** [comment]
@kounelisagis this looks good -- a re-run of what I posted above:

```
>>> A = tiledb.open('snda')
```

This is now a `property` -- thank you!

```
>>> cd = A.schema.current_domain
```

This is the default repr -- I'd love it _as a bonus_ if this could print something useful, but, this PR is fine without:
```
>>> cd
<tiledb.current_domain.CurrentDomain object at 0x71d099408680>
```

```
>>> cd. <----- TAB-COMPLETION
cd.from_capsule(     cd.is_empty          cd.set_ndrectangle(
cd.from_pybind11(    cd.ndrectangle       cd.type
```

These look good:

```
>>> cd.is_empty
False

>>> cd.type
<CurrentDomainType.NDRECTANGLE: 0>

>>> nd = cd.ndrectangle
```

Again no non-default repr but this is OK:

```
>>> nd
<tiledb.ndrectangle.NDRectangle object at 0x71d06dba4400>

>>> nd. <----- TAB-COMPLETION
nd.from_capsule(   nd.from_pybind11(  nd.range(          nd.set_range(
```

These look good:

```
>>> nd.range(0)
(0, 99)

>>> nd.range(1)
(0, 199)

>>> nd.range('soma_dim_0')
(0, 99)

>>> nd.range('soma_dim_1')
(0, 199)
```

These are nicely informative:

```
>>> nd.range(2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/git/TileDB-Inc/TileDB-Py/tiledb/ndrectangle.py", line 61, in range
    return self._range(dim, self._domain.dim(dim).dtype)
  File "/home/ubuntu/git/TileDB-Inc/TileDB-Py/tiledb/domain.py", line 172, in dim
    return Dim.from_pybind11(self._ctx, self._dim(dim_id))
tiledb.cc.TileDBError: C API: Dimension index 2 is out of bounds; valid indices are 0 to 1

>>> nd.range('soma_data')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/git/TileDB-Inc/TileDB-Py/tiledb/ndrectangle.py", line 61, in range
    return self._range(dim, self._domain.dim(dim).dtype)
  File "/home/ubuntu/git/TileDB-Inc/TileDB-Py/tiledb/domain.py", line 172, in dim
    return Dim.from_pybind11(self._ctx, self._dim(dim_id))
tiledb.cc.TileDBError: C API: Dimension 'soma_data' does not exist
>>> nd.range('nonesuch')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/git/TileDB-Inc/TileDB-Py/tiledb/ndrectangle.py", line 61, in range
    return self._range(dim, self._domain.dim(dim).dtype)
  File "/home/ubuntu/git/TileDB-Inc/TileDB-Py/tiledb/domain.py", line 172, in dim
    return Dim.from_pybind11(self._ctx, self._dim(dim_id))
tiledb.cc.TileDBError: C API: Dimension 'nonesuch' does not exist
```

It's a _little_ weird we can't ask the `cd` or the `nd` for its number of slots but this is OK -- we have the schema:

```
>>> for i in range(A.schema.ndim): print(nd.range(i))
...
(0, 99)
(0, 199)
```

**Review by johnkerl — APPROVED** [review]
@kounelisagis thank you!

For completeness we can have the core datetime types as well but we can do that as a follow-on PR

**Pull request merged**